docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,196 @@
+6. Then we see obj3, and s0_1 indicates that it should be deduped to s0 so we change obj3’s ref to point to s0_1 right away.
+7. Then we see obj4 which points to s1 and s1 indicates it’s stored in the hash table so we don’t need to dedup.
+
+Since we know the size of a string object trivially, we know which strings are on LOH.
+
+**Situations when we choose to not dedup**
+
+- If `InterlockCompareExchangePointer` fails because the ref was modified while we were finding a copy to dedup to (or insert into the hash table), we skip this ref.
+- If too many collisions exist for a hash code, we skip deduping for strings with that hash code. This avoids the DoS attack by creating too many strings for the same hash.
+
+**Alternate design points**
+
+- Should we calculate the hash codes for SOH strings as gen1 GCs promote them into gen2? 
+
+This would increase gen1 pause.
+
+- Should we dedup while creating LOH strings?
+
+This would reduce LOH string allocation speed and potentially by a lot.
+
+#### **How often to dedup**
+
+If there’s no new objects that get into gen2/LOH, there’s no need to run dedup as the duplicates would have already been eliminated (mostly) during the last dedup cycle. So we should use the new amount of objects these generations get to determine how often to dedup, ie, taking the following factors into consideration -
+
+- We know when new LOH strings are allocated (all LOH allocations are done via C++ helpers in gchelpers.cpp), we can take the accumulative size of new LOH strings till it reaches a certain percentage of LOH before starting the next dedup cycle.
+
+- We know how much is promoted into gen2 and we can also wait till a certain percentage of gen2 consists of newly promoted memory before starting the next dedup cycle.
+
+#### **Compatibility**
+
+The following scenarios become problematic or more problematic when deduping is on so we need ways to help users find them.
+
+- Mutating the string content
+
+Strings are supposed to be immutable. However in unsafe code you can change the string content after it’s created. Changing string content already asking for trouble without deduping – you could be changing the interned copy which means you are modifying someone else’s string which could cause completely unpredictable results for them. 
+
+The most common way is to use the fixed keyword:
+
+`fixed (char* p = str)`
+
+`{`
+
+  `p[0] = 'C';`
+
+`}`
+
+There are other ways such as 
+
+`((char*)(gcHandlePointingToString.AddrOfPinnedObject())) = 'c';`
+
+Or
+
+`Unsafe.AsRef (in str.GetPinnableReference()) = 'c';`
+
+- Locking on a string
+
+Locking on a string object is already discouraged due to a string can be interned. Having string dedup on can make this problematic more often if the string you called lock on is now deduped to a different string object. 
+
+- Reference equality
```

Comment:
You can include this example from https://github.com/dotnet/coreclr/pull/15135#issuecomment-349117182 to illustrate the problem:

```
    string _field;

    void Test()
    {
        _field = new string('a', 1);

        string a1 = _field;
        Thread.Sleep(1000);
        string a2 = _field;

        // This assert can fail with string deduping turned on
        Debug.Assert(ReferenceEquals(a1,a2));
    }
}
```

Reply:
Is this example correct? It seems that given strings _a_ and _b_ they can go from `!ReferenceEquals` to `ReferenceEquals` but not the other way around. Perhaps I'm misreading the propsoal?

Reply:
If the pointer `a1` on the stack hasn't been deduped; however the pointer `_field` has been deduped then `.ReferenceEquals(a1, _field)` will be false?

Reply:
As @benaadams said.

Reply:
will add this example.